File: src/main/java/com/flowcommerce/core/storage/BaseEntity.java
package com.flowcommerce.core.storage;

import jakarta.persistence.*;
import lombok.Getter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

@Getter
@MappedSuperclass
public abstract class BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private EntityStatus status = EntityStatus.ACTIVE;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    public void active(){
        this.status = EntityStatus.ACTIVE;
    }

    public Boolean isActive() {
        return status == EntityStatus.ACTIVE;
    }

    public void delete(){
        this.status = EntityStatus.DELETED;
    }

    public Boolean isDeleted() {
        return status == EntityStatus.DELETED;
    }

}

File: src/main/kotlin/com/flowcommerce/core/storage/BaseEntity.kt
package com.flowcommerce.core.storage

import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.LocalDateTime

@MappedSuperclass
abstract class BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0
        protected set

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: EntityStatus = EntityStatus.ACTIVE
        protected set

    @CreationTimestamp
    var createdAt: LocalDateTime? = null
        protected set

    @UpdateTimestamp
    var updatedAt: LocalDateTime? = null
        protected set

    fun active() {
        this.status = EntityStatus.ACTIVE
    }

    val isActive: Boolean
        get() = status == EntityStatus.ACTIVE

    fun delete() {
        this.status = EntityStatus.DELETED
    }

    val isDeleted: Boolean
        get() = status == EntityStatus.DELETED
}
